### PR TITLE
Provide basic support for #fragment section

### PR DESCRIPTION
**Pull request by angelozerr**
Provide basic support for #fragment section

image

This PR provides 
 * `#fragment` snippet  completion
 * avoid reporting undefined section tag for #fragment
 * improve completion and hover documentation fo all sections by adding links at the end of the documentation

See https://github.com/redhat-developer/vscode-quarkus/issues/563

Signed-off-by: azerr <azerr@redhat.com>

### CODE REVIEW

**Comment by angelozerr** [comment]
@datho7561 I had had a strange problem. When I execute test on my local machine, I have some trouble (although CI build seems working?) I had to fix https://github.com/redhat-developer/quarkus-ls/pull/768/files#diff-0c6cba217b8d2aafec174ac4005b6f93e7e446f63641556079e995424cb62330

**Comment by angelozerr** [comment]
For fragment syntax coloration, see https://github.com/redhat-developer/vscode-quarkus/pull/570

**Comment by datho7561** [comment]
Locally, I have a compilation issue, but I think it's just that I installed a version of lsp4mp that hasn't been published

File: qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/ls/commons/snippets/Link.java [hunk 1]
```diff
@@ -0,0 +1,38 @@
+/*******************************************************************************
+* Copyright (c) 2020 Red Hat Inc. and others.
```
Comment (datho7561) [inline qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/ls/commons/snippets/Link.java:2]: 2023

**Comment by angelozerr** [comment]
I have the impression that CI build doesnt exe ute QuteHoverInUserTag?

**Comment by angelozerr** [comment]
If you execute  QuteHoverInUserTag.java in master you have 2 errors,No?

File: qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/parser/template/sections/FragmentSection.java [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+
+	public static final String TAG = "fragment";
+
+	private static final String ID = "id";
+
+	private static final String RENDERED = "rendered";
+
+	private static final ParametersInfo PARAMETER_INFOS = ParametersInfo.builder() //
+			.addParameter(ID) //
+			.addParameter(new ParameterInfo(RENDERED, null, true)) //
```
Comment (datho7561) [inline qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/parser/template/sections/FragmentSection.java:51]: I don't seem to have completion/validation for the parameters (except for `id` being added when I complete the snippet). Is there supposed to be completion/validation? If not, maybe we could look into implementing it (in a different PR).

File: qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/utils/DocumentationUtils.java [hunk 3]
```diff
@@ -340,14 +344,9 @@ private static void addUrl(String url, StringBuilder documentation, boolean mark
...
 			documentation.append("See ");
-			if (markdown) {
-				documentation.append("[here](");
-				documentation.append(url);
-				documentation.append(")");
-			} else {
-				documentation.append(url);
-			}
+			addLink(url, "here", documentation, markdown);
 			documentation.append(" for more informations.");
```
Comment (datho7561) [inline qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/utils/DocumentationUtils.java:348]: This should be: `" for more information."` (i.e. no `s` at the end of information).

File: qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/ls/commons/snippets/SnippetRegistry.java [hunk 4]
```diff
@@ -315,9 +315,42 @@ private static MarkupContent createDocumentation(Snippet snippet, Map<String, St
...
+				if (i > 0) {
+					if (i == links.size() - 1) {
+						documentation.append(" and ");
+					} else {
+						documentation.append(", ");
+					}
+				}
+				addLink(link.getUrl(), link.getLabel(), documentation, canSupportMarkdown);
+			}
+			documentation.append(" for more informations.");
```
Comment (datho7561) [inline qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/ls/commons/snippets/SnippetRegistry.java:338]: Same here: `" for more information."`

**Review by datho7561 — CHANGES_REQUESTED** [review]
Functionality seems to be working pretty good, although I found some things in the code.

**Comment by datho7561** [comment]
Wait @angelozerr, do you have test failures with the code that is pushed? I don't have any.

**Comment by angelozerr** [comment]
> Wait @angelozerr, do you have test failures with the code that is pushed? I don't have any.

If I took master, tests fails and I fix with https://github.com/redhat-developer/quarkus-ls/pull/768/files#diff-0c6cba217b8d2aafec174ac4005b6f93e7e446f63641556079e995424cb62330 which is out of the scope of this PR.

**Comment by datho7561** [comment]
If I check out `master`, the tests aren't failing for me locally. Given that the solution was updating a dependency, maybe it has something to do Maven repo configuration?

File: qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/ls/commons/snippets/Link.java (hunk 1 above)
Comment (angelozerr) [reply qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/ls/commons/snippets/Link.java]: fixed

File: qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/utils/DocumentationUtils.java (hunk 3 above)
Comment (angelozerr) [reply qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/utils/DocumentationUtils.java:348]: fixed

File: qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/ls/commons/snippets/SnippetRegistry.java (hunk 4 above)
Comment (angelozerr) [reply qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/ls/commons/snippets/SnippetRegistry.java]: fixed

Comment (angelozerr) [reply qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/ls/commons/snippets/SnippetRegistry.java]: I changed the render  of links when there are several urls, I display then with bullet:

image

**Comment by angelozerr** [comment]
> If I check out master, the tests aren't failing for me locally. Given that the solution was updating a dependency, maybe it has something to do Maven repo configuration?

The 2 tests fails because those tests are wrong, the documentation should show Javadoc from field. In master the Javadoc is not included. It is my fix.

File: qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/parser/template/sections/FragmentSection.java (hunk 2 above)
Comment (angelozerr) [reply qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/parser/template/sections/FragmentSection.java:51]: I agree with you, we need this kind of support. Today ParameterInfos are used only to know if a parameter can be assigned to an expression or if the parameter is an expression.

I think it should be nice to use it for completion and validation (with a generic mean). Let's see that in an another PR.

**Review by datho7561 — APPROVED** [review]
Looks good and works well. Thanks, Angelo!

**Pull request merged**